### ISSUE TITLE
Debug Capacitor Ionic/Angular in Browser

### ISSUE THREAD

**Issue by mikkimichaelis** [ionic-team/vscode-extension#86]
I've installed the vscode Ionic extension expecting to see some debug options as they appear in documentation but I have none.  Any reason why?  

Why is a capacitor app so difficult to debug in vscode while running in a browser?  As awesome as capacitor seems to be, the loss of setting breakpoints in vscode is crippling.

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis,

The extension supports web and Android debugging. I'm taking a look at an issue where launching `Debug` > `Web` is not launching the debugger (will update when that is fixed). If you are looking into Android Debugging I've written up some troubleshooting info below.

**Android Debugging**
The extension relies on Android Studio to be installed on the same machine and it will use ADB (Android Debug Bridge) to debug applications and allow you to set breakpoints etc. ADB comes with Android Studio but if your machine was not installed with it you can get as part of the [Platform Tools](https://developer.android.com/studio/releases/platform-tools.html).

If the extension can't find ADB then it won't show the Debug option on the side menu. Occasionally, installs of Android Studio will be placed where the extension cannot find them. To fix this you can go to `Settings` > `Advanced` and specify the `Adb Path` . The typical folder where it is found is `/Users/{your username}/Library/Android/sdk/platform-tools`

In Android Studio you can verify you have tools installed by going to Tools > SDK Manager > SDK Tools and ensure you have tools install for the Android versions you want to debug.

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis,

Web debugging should now be fixed in v1.22.1. The debugger was not launching if your project was Angular based and now all you have to do is click `Debug` > `Web` and it will compile, launch a debuggable browser and then start a VS Code debugging session.

One more important note: If your Angular build is set to `Production` then you may not be able to debug (because your Angular configuration is not outputting source maps), You may need to click the 3 dots next to `Build` and choose `Development`.

**Comment by mikkimichaelis** [comment]
wow, how exciting!  I just checked the plugin and it was requesting an update from v1.22.0 -> v1.22.1 this morning!  It just upgraded and I'm very sad to say, there is no Debug folder.  

I am using Angular as my framework.  How can I verify my Angular build is Development?  Honestly I'm really not even sure where my map files are being generated when I run

`ionic build`
or
`ng build`

**Comment by mikkimichaelis** [comment]
Any update on this issue?  I would think it would be a rather high priority one.  Maybe I'm doing something wrong, I can't imagine Ionic/Capacitor still being a viable framework in modern times without the ability to debug via web browser.  I must be doing something wrong, please help.

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis ,

Can you try a couple of things: First, open a new folder and create a new Angular project, add the Android Integration. You should see an item called `Debug`.  If that works, then we know its your particular project that has something about it not working and if you are able to share your project then I can find out what the problem is and resolve it.

**Comment by mikkimichaelis** [comment]
hi @dtarnawsky and thank you for your reply.  I did as you suggested and generated a new ionic/angular project and from within vscode the Ioinic extension has a Debug web option.  Excellent!

Therefore the issue must be with my existing project.  I have cut everything out of my project and published a repo here https://github.com/mikkimichaelis/ionic-vscode-test.git

This repo does not have the Debug option when running.  My first guess as to the problem may be that I use a custom webpack.config (that's all but empty for development) but may still be causing the issue.  This is a rather old project that has gone through many angular and ionic upgrades over the past few years.  Thank you so much for your help!

Mikki

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis ,

You've got a **Cordova** project!

If you want to debug you'll need to follow the advice of the recommendation the extension gives of "Remove Cordova"

This will remove the `config.xml` file and the `cordova` section of your `package.json`

After this is done the extension will show all the extra Capacitor features like debug.

However, there is still a problem with your @angular-builders/custom-webpack
I tried creating a configuration for development by adding:
```
                        "development": {
                            "buildOptimizer": false,
                            "optimization": false,
                            "vendorChunk": true,
                            "extractLicenses": false,
                            "sourceMap": true,
                            "namedChunks": true
                          },  
```

The sourceMap being true is what you need for debugging.

But your build fails with:
```
[ng] Error: Schema validation failed with the following errors:
[ng]   Data path "" must NOT have additional properties(sourceMap).
```

If sourceMaps can be generated then you can debug. Also, if you ditch the custom-webpack I'm sure you can debug as well.

**Comment by mikkimichaelis** [comment]
Does this mean that I must remove all Cordova plugins to be able to web debug?  Obviously none of my Cordova plugins are being used in the web app, so I will try to eliminate the Cordova integration only on the web branch and see how that goes.  Please let me know if this is correct.  Thanks again!

Also, watching the video at https://vs-ionic.netlify.app/videos/cap-migration.gif referenced in the vscode plugin docs I saw the removal of all plugins and Cordova itself, but at the end of the video the Debug option still did not show up as I was somewhat expecting.  I'm assuming a plugin restart would be required?

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis,

You **do not** need to remove any Cordova plugins (unless you know they are part of [the few](https://capacitorjs.com/docs/plugins/cordova#known-incompatible-plugins) that don't work with Capacitor).

You only need to remove `config.xml` and the `cordova` section in package.json - these are the remnants of when your project was Cordova only and aren't referenced by Capacitor at all, so they can be deleted.

With the video, we'll have some new videos coming out soon (and more docs) to help explain things a little better.

Also, thanks so much for posting up your repo - it helps a lot to see what kind of variations of projects people can come up with which tests the edge cases of the extension.

**Comment by mikkimichaelis** [comment]
ok, trying to remove the Cordova plugins and Cordova itself I receive the following error

 env: node: No such file or directory

I'm assuming the ionic plugin is looking for an env variable or something because node is most definitely in my path.

**Comment by mikkimichaelis** [comment]
This is odd, I just returned to the vscode project and it seems now the remove Capacitor Plugin and Remove Capacitor commands did work as I see modification to the project and these are no longer shown under Recommendations.   Cordova.xml was renamed to .bak, 'corodva' section removed from package.json, and 'cordova' removed from ionic.config.json integrations section (however 'capacitor' remains as an integration).

From within the ionic vscode plugin the only suggestions are now 'Add Capacitor Integration' and 'codelyzer'.  When clicking Add Capacitor I receive the following in output:

[Ionic] Add Capacitor...
> [@ionic-init]
[Ionic] Creating Ionic project...
[Ionic] Created Ionic Project
Capacitor added to this project
[Ionic] Add Capacitor Integration Completed.

[Error] env: node: No such file or directory

[Error] env: node: No such file or directory

...which are essentially the same errors I previously received and yet the commands properly completed (I think).  I've restarted vscode and yet I still do not see the option to Debug.  

Thanks again for all ur help!

**Comment by mikkimichaelis** [comment]
Ok, I've resolved the issue and now have the Debug option!  I was receiving the above error [Error] env: node: No such file or directory basically trying to do anything with the ionic plugin, researched it up and found this post [https://github.com/jest-community/vscode-jest/issues/105](https://github.com/jest-community/vscode-jest/issues/105) suggesting to launch vscode within the root of the project using 'code .' Previously I've been launching vscode from my mac launcher and then opening the project directory.  This resolved the node error.  I hope this little bit helps another developer.

Now, selecting the Debug -> Web option shows some warnings and errors as seen below (which seem to be non critical) but then goes into an endless wait for npm (as does running the command line 'ionic serve' command, however the site does build, launch, and run as expected.

[npm] npm WARN config cache-min This option has been deprecated in favor of `--prefer-offline`.
[Error] npm WARN config cache-min This option has been deprecated in favor of `--prefer-offline`.
[INFO] Waiting for connectivity with npm...
[INFO] Waiting for connectivity with npm...
[INFO] Waiting for connectivity with npm...
etc

I am unable to set breakpoints, nor does adding a debugger statement to the code stop execution with chrome dev tools open (weird).

I have pushed all these updates to [https://github.com/mikkimichaelis/ionic-vscode-test](https://github.com/mikkimichaelis/ionic-vscode-test)

Any help getting breakpoints to be hit, or understanding the waiting for npm would be greatly appreciated!  Thank you so much!

UPDATE: I fixed the cache-min error/warning by editing .npmrc and removing the reference to cache-min.  I also added 

```
prefer-offline: true
```

I was receiving the ‘Waiting for connectivity with npm..’ because I had at some point added custom ‘ionic:build’ and ‘ionicse scripts to package.json.

```
"ionic:build": "npm run build",
"ionic:serve": "npm run start”
```

I had at some point in the journey of this app from Ionic 3 to Ionic 6 added these scripts for a legitimate reason.  Removing them fixed connectivity with npm.

**Comment by mikkimichaelis** [comment]
Ok, I revisited this issue today and fixed the following issues.  The first issue was that through the angular upgrades the entire “development” section was removed from angular.json in two places (or was never there in the first place).  Adding the following section to “build -> configurations” began emitting sourceMaps

```
"development": {
    "buildOptimizer": false,
    "optimization": false,
    "vendorChunk": true,
    "extractLicenses": false,
    "sourceMap": true,
    "namedChunks": true
},
```
and in the “serve -> configuration” section of angular.json I added

```
"development": {
    "browserTarget": "app:build:development”
  },
},
"defaultConfiguration": “development"
```

Next was dealing with webpack, searching google I found I need to add the following to my webpack.config.js

```
devtool: 'source-map'
```

With these changes I am now able to debug my capacitor app in the browser.  I still don’t have the options to debug Android or iOS even though both those platforms are in the project working fine, and there are no plugin capacitor recommendations to add them.  another day.

**Comment by mikkimichaelis** [comment]
Good grief!  Ok, so yes, I can set  and hit breakpoints in any of my pages and components, but not in any pipes, guards, services, or even app.component.ts itself.


I deleted all webpack references and reverted back to the following as necessary within angular.json.  Project builds and runs just fine.  Debugging is another story.

```
"builder": "@angular-devkit/build-angular:browser”,
"builder": "@angular-devkit/build-angular:dev-server”,
```
...made no change in debugging capabilities.

**Comment by mikkimichaelis** [comment]
TLDR;
Breakpoints in a new Ionic/Angular/Capacitor/Blank starter project with latest (and prerelease) ionic vscode plugin are not being bound or hit.  debugger statements are being compiled out.  please help.

...
I decided to start with a new ionic project and test its debugging capabilities.  running

```
npm i @ionic/cli@latest -g
ionic --version
```

I receive the result

v6.20.8

next running

```
ionic start ionic-debug-test
```
selecting Angular
selecting blank starter

a new ionic/angular/capacitor app is created from the ‘blank’ template

```
cd ionic-debug-test
code .
```

I edited home.page.ts and added the following:

```node
constructor() {
    debugger;
    console.log(‘wtf')
  }
```

I set a breakpoint on the console.log() statement, expecting at least one of them to be hit.

from the Ionic plugin I selected Debug -> Web and the build started, completed and launched a browser.

‘wtf’ was logged to the console., no breakpoint was hit.  examining the breakpoint, it was unbound and greyed out.  Running 

```
ionic build
```
and examining the resulting .js files, the debug statement was no where to be found.  Also, no .map files were generated in the www directory.

I repeated this process, creating a new ionic app from a template but this time selecting the React framework.

Long story short, the breakpoints work just fine with React.

I noticed there is a prerelease version of the Ionic plugin available, so returning to the Angular app and I updated.  Breakpoints still can not be bound in the angular project.

Two weeks ago I commented that a new ionic/angular/capacitor app was working, however I assumed it was working because the Debug -> Web option was now available within the project.  However I made the mistake in not actually trying to debug the project, as I see now it was not actually working.

It would be greatly appreciated if somebody would help me understand why this is not working.

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis ,

For an Angular 15 project, Angular will default to a production build if no configuration is specified. So to debug you would need to select the `...` next to `build` and choose `development`. After this when you debug breakpoints will be hit.

Try this out and I'm sure it will work.

As this may not be obvious I'll see what I can do to automatically switch to "development" for you when you click debug.

**Comment by mikkimichaelis** [comment]
Awesome @dtarnawsky  

OMG THANK YOU SO MUCH!!!!!!  Seriously, send me ur Buy Me a Coffee, or whatever (mikkimichaelis@gmail.com), so I can buy you a coffee or beer, please!  I’ve been going at it so hard the past few months using console.log() to debug, your help has literally change my world!

I can now set and hit breakpoints in my entire app!  Was so weird and frustrating only being able to set breakpoints in pages.

Also, I’d considered that a production build was being created and thus set 

```
"defaultConfiguration": “development"
```

in both the architect->build & serve sections of angular.json.  I should have investigated the settings of the vsccode plugin more throughly.   I also had to enable HTTPS via the plugin because my authentication system requires it.

I am so happy, this progress has literally made my day!  I greatly appreciate all your hard work :-)

I hope this Issue now being resolved helps maintain the sanity of some other poor unfortunate soul.

FYI - I am still using the angular custom web pack build, with a web pack config that only consists of 

```node
module.exports = {
    mode:’development’
}
```

**Comment by dtarnawsky** [comment]
Hi @mikkimichaelis ,  just to follow up on this. v1.24.0 has been released and whenever you click the `Debug` button for web it will check if you have `production` as your configuration and temporarily switch it to `development`. That way you don't have to remember to change your configuration before debugging.

Glad you have your debugging going!
